common/network-types/src/session/segmenter.rs

```diff
@@ -0,0 +1,319 @@
+use futures::channel::mpsc::SendError;
+use futures::{ready, Sink};
+use pin_project::pin_project;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+use crate::prelude::protocol::SessionMessage;
+use crate::session::frame::FrameId;
+use crate::session::frame::SeqNum;
+use crate::session::Segment;
+
+#[pin_project]
+pub struct Segmenter<const C: usize> {
+    frame_size: usize,
+    seg_buffer: Vec<u8>,
+    ready_segments: Vec<Segment>,
+    next_frame_id: FrameId,
+    seq_len_bytes: usize,
+    closed: bool,
+    #[pin]
+    tx: futures::channel::mpsc::Sender<Segment>,
+}
+
+impl<const C: usize> Segmenter<C> {
+    const PAYLOAD_CAPACITY: usize = C - SessionMessage::<C>::SEGMENT_OVERHEAD;
+
+    pub fn new(frame_size: usize, capacity: usize) -> (Self, impl futures::Stream<Item = Segment> + Send) {
+        assert!(frame_size >= C, "frame size must be at least MTU");
+        assert!(
+            frame_size <= C * SessionMessage::<C>::MAX_SEGMENTS_PER_FRAME,
+            "frame size too big for the given MTU"
+        );
```

Comment:
Should then frame_size be a compile time arg passed to new to be conditioned upon?

Reply:
Can be